### PR TITLE
[analyzer] Harden safeguards for Z3 query times

### PR DESCRIPTION
**Pull request by steakhal**
This patch is a functional change.
https://discourse.llvm.org/t/analyzer-rfc-taming-z3-query-times/79520

As a result of this patch, individual Z3 queries in refutation will be
bound by 300ms. Every report equivalence class will be processed in
at most 1 second.

The heuristic should have only really marginal observable impact -
except for the cases when we had big report eqclasses with long-running
(15s) Z3 queries, where previously CSA effectively halted.
After this patch, CSA will tackle such extreme cases as well.

### CODE REVIEW

**Comment by steakhal** [comment]
I've tested this change on 200+ projects, and the overall effect is:
 - 0 new issues appear
 - 276 issues disappear (because we drop the report eqclass)
 
 The three most affected checker categories (all of them are spread across usually 20+ projects):
  - null-deref-like diagnostics (74)
 - OOBv2 (70)
 - uninit-read-like (55)
 
 This underpins the non-intrusive nature of this change.

**Comment by steakhal** [comment]
Ping. @NagyDonat @Xazax-hun @haoNoQ

File: clang/include/clang/StaticAnalyzer/Core/BugReporter/Z3CrosscheckVisitor.h [hunk 1]
```diff
@@ -44,21 +47,43 @@ class Z3CrosscheckVisitor final : public BugReporterVisitor {
...
+  /// - Hit the 400'000 rlimit in the report eqclass.
+  ///
+  /// Refer to
+  /// https://discourse.llvm.org/t/analyzer-rfc-taming-z3-query-times/79520 to
+  /// see why this heuristic was chosen.
+  Z3Decision interpretQueryResult(const Z3CrosscheckVisitor::Z3Result &Meta);
+
+private:
+  const AnalyzerOptions &Opts;
+  unsigned NumZ3QueriesDoneInEqClass = 0;
```
Comment (steakhal) [inline clang/include/clang/StaticAnalyzer/Core/BugReporter/Z3CrosscheckVisitor.h:85]: Hmm, I dont think I use this anymore.

File: clang/lib/StaticAnalyzer/Core/Z3CrosscheckVisitor.cpp [hunk 2]
```diff
@@ -101,18 +123,37 @@ void Z3CrosscheckVisitor::Profile(llvm::FoldingSetNodeID &ID) const {
...
     const Z3CrosscheckVisitor::Z3Result &Query) {
   ++NumZ3QueriesDone;
+  ++NumZ3QueriesDoneInEqClass;
+  AccumulatedZ3QueryTimeInEqClass += Query.Z3QueryTimeMilliseconds;
 
-  if (!Query.IsSAT.has_value()) {
-    // For backward compatibility, let's accept the first timeout.
+  if (Query.IsSAT && Query.IsSAT.value()) {
+    ++NumTimesZ3QueryAcceptsReport;
+    return AcceptReport; // sat
```
Comment (NagyDonat) [inline clang/lib/StaticAnalyzer/Core/Z3CrosscheckVisitor.cpp:131]:
```suggestion
    return AcceptReport;
```
This old comment doesn't add anything + the "unset" on the other branch was slightly incorrect.

File: clang/unittests/StaticAnalyzer/Z3CrosscheckOracleTest.cpp [hunk 3]
```diff
@@ -17,43 +18,126 @@ using Z3Decision = Z3CrosscheckOracle::Z3Decision;
...
+  // 35*20ms = 700ms, which is equal to the 700ms threshold.
+  for (int i = 0; i < 35; ++i) {
+    ASSERT_EQ(RejectReport, interpretQueryResult({UNSAT, 20_ms, 1000_step}));
+  }
+  // Do one more to trigger the heuristic, but given this was SAT, we still
+  // accept the query.
+  ASSERT_EQ(AcceptReport, interpretQueryResult({SAT, 200_ms, 1000_step}));
+}
+
+TEST_F(Z3CrosscheckOracleTest, RejectEQClassIfHitsRLimitTooManyTimes) {
```
Comment (NagyDonat) [inline clang/unittests/StaticAnalyzer/Z3CrosscheckOracleTest.cpp:131]: The name `"HitsRLimitTooManyTimes"` is a leftover from the earlier plan where one `rlimit` overrun wasn't enough to discard the eqclass. (This also applies for the next TC.)

**Review by NagyDonat — APPROVED** [review]
Oops, I vaguely recalled that I already reviewed this, but it turns out I didn't. Anyway, the change LGTM, I only added a few minor remarks. (We already discussed a high-level overview of this on Discourse.)

**Comment by steakhal** [comment]
Ah, it mixed the messages... Here it is:

- Removed unused `NumZ3QueriesDoneInEqClass`
- Added a new `crosscheck-with-z3-eqclass-timeout-threshold` config option.
- Removed misleading `// unsat` comment on return statement.
- Removed unnecessary `// sat` comment on return statement.
- Adjusted the name of the two last unit-tests.

**Comment by steakhal** [comment]
Reverting this patch as requested in this [comment](https://github.com/llvm/llvm-project/pull/95128#issuecomment-2176008007).

**Pull request merged**